src/devices/machine/tk10_keyboard.cpp

```diff
@@ -0,0 +1,256 @@
+		return (BIT(kbspecial, 0) & BIT(kbspecial, 1));
+	});
+
+	m_tk10cpu->t0_in_cb().set([this] () -> u8
+	{
+		return (BIT(m_kbspecial->read(), 2));  // control key
+	});
+
+	m_tk10cpu->p1_in_cb().set([this] () -> u8
+	{
+		for (int i = 0; i < 8; i++)
+		{
+			if (!BIT(m_tk10cpu->p2_r(), i)) return m_keyboard[i]->read();
+			// read key matrix:
+			// a low value in p2 bit will drive bit of p1 low when switch contact made
+		}
+		return 0xff;
+	});
+
+	m_tk10cpu->bus_out_cb().set([this] (u8 data)
+	{
+		m_data_write(data); // callback with data in low 7 bits and strobe in msb
+	});
+}
+
+void tk10_keyboard_device::tk10_mem_map(address_map &map)
+{
+	map(0x0000, 0x07ff).rom().region("tk10", 0);
+}
+
+INPUT_CHANGED_MEMBER(tk10_keyboard_device::tk10_func)
+{
+		m_tk10cpu->set_input_line(MCS48_INPUT_IRQ, newval ? CLEAR_LINE : ASSERT_LINE);
+}
+
+INPUT_CHANGED_MEMBER(tk10_keyboard_device::tk10_reset)
+{
+	// printf("TK10 RESET new=%x old=%x\n",newval,oldval);
+	u8 key = m_kbspecial->read();
+	if (!BIT(key, 3) && !BIT(key, 4)) m_reset_write(0);
+	else m_reset_write(1);
+}
+
+
+
+//**************************************************************************
+//  INPUT PORTS
+//**************************************************************************
+
+static INPUT_PORTS_START(tk10_keyboard)
+
+	PORT_START("P20")
+	PORT_BIT(0x001, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Esc") PORT_CODE(KEYCODE_ESC)      PORT_CHAR(27)
+	PORT_BIT(0x002, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Edit Function Key") PORT_CODE(KEYCODE_HOME)
+	PORT_BIT(0x004, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("End Edit Function Key") PORT_CODE(KEYCODE_END)
+	PORT_BIT(0x008, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Insert Text") PORT_CODE(KEYCODE_INSERT)  // moves up, now inserts text, enter returns
+	PORT_BIT(0x010, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_0_PAD)   // CALL -151
+	PORT_BIT(0x020, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Pause Key") PORT_CODE(KEYCODE_UP)  // CTRL+S 0x13 PAUSE OUTPUT
+	PORT_BIT(0x040, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_9_PAD)   // UNDERSCORE   7F
+	PORT_BIT(0x080, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("TAB") PORT_CODE(KEYCODE_8_PAD)   // CTRL+i 0x09  tab??
+
+	PORT_START("P21")
+	PORT_BIT(0x001, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_ENTER_PAD)   //  _ SHIFT=^
+	PORT_BIT(0x002, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_PGUP)   // ]
+	PORT_BIT(0x004, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_CLOSEBRACE)  // [
+	PORT_BIT(0x008, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_PGDN)        // @ SHIFT=^
+	PORT_BIT(0x010, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_SLASH_PAD)   // /
+	PORT_BIT(0x020, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_MINUS_PAD)
+	PORT_BIT(0x040, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_PLUS_PAD)
+	PORT_BIT(0x080, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_ASTERISK)
+
+	PORT_START("P22")
+	PORT_BIT(0x001, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_BACKSLASH) // BACKSLASH
```

Comment:
Do these keys really have blank key caps?  It’s bad to have inputs that lack either `PORT_NAME` or `PORT_CHAR`.

Reply:
ok, tried to label them all